image/tarexport/save_oci.go

```diff
@@ -0,0 +1,309 @@
+package tarexport
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"regexp"
+
+	oci "github.com/containers/image/oci/layout"
+	ctypes "github.com/containers/image/types"
+	"github.com/docker/distribution/digest"
+	"github.com/docker/docker/image"
+	"github.com/docker/docker/layer"
+	"github.com/docker/docker/pkg/archive"
+	"github.com/docker/docker/reference"
+	imgspec "github.com/opencontainers/image-spec/specs-go"
+	imgspecv1 "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+type layerInfo struct {
+	digest string
+	size   int64
+}
+
+type ociSaveSession struct {
+	*ocitarexporter
+	// string is a tag here
+	images       map[string]image.ID
+	refs         map[string]string
+	name         string
+	savedImages  map[image.ID][]byte // cache image.ID -> manifest bytes
+	diffIDsCache map[layer.DiffID]*layerInfo
+}
+
+func (l *ocitarexporter) Save(names []string, outStream io.Writer) error {
+	images, err := l.parseNames(names)
+	if err != nil {
+		return err
+	}
+
+	return (&ociSaveSession{ocitarexporter: l, refs: l.refs, images: images}).save(outStream)
+}
+
+var ociRefRegexp = regexp.MustCompile(`^([A-Za-z0-9._-]+)+$`)
+
+func (l *ocitarexporter) parseNames(names []string) (map[string]image.ID, error) {
+	tags := make(map[string]image.ID)
+
+	addAssoc := func(name string, id image.ID, ref reference.Named) error {
+		if ref == nil {
+			if r, ok := l.refs[name]; ok {
+				if !ociRefRegexp.MatchString(r) {
+					return fmt.Errorf(`invalid reference %q, reference must not include characters outside of the set of "A" to "Z", "a" to "z", "0" to "9", the hyphen "-", the dot ".", and the underscore "_"`, r)
+				}
+				tags[r] = id
+			} else {
+				if !ociRefRegexp.MatchString(name) {
+					return fmt.Errorf("invalid reference %q, please specify a reference with --ref with a valid characters set", name)
+				}
+				tags[name] = id
+			}
+		}
+
+		if ref != nil {
+			var (
+				tagged reference.NamedTagged
+				err    error
+				ociRef string
+			)
+			if c, ok := ref.(reference.Canonical); ok {
+				// using the digest's hex as reference when saving name@digest, discuss...
+				ociRef = c.Digest().Hex()
```

Comment:
@stevvooe that hex you had under `./refs/` in https://github.com/docker/docker/pull/26369#issuecomment-249333719 probably comes from here. You're saving w/o a tag or digest and so when saving you get all the references to that image name - thus, you had a digest in you docker images and that got saved using the hex as the ref.